### PR TITLE
Feature/ab#12 users controller

### PR DESCRIPTION
**Pull request by BerendWouters**
This includes a basic working example of the CareUser table and a serverside pagination mechanism.

### CODE REVIEW

File: Singer.API/Models/SearchResults.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+         Items = items;
+         TotalCount = totalItemsCount;
+         Start = pageIndex;
+         Size = items.Count;
+      }
+
+      public int TotalCount { get;  }
+      public int Start { get;  }
+      public int Size { get; }
+      public IReadOnlyList<T> Items { get; }
```
Comment (wim07101993) [inline Singer.API/Models/SearchResults.cs:23]: Why is it an IReadOnlyList here while in the pagination model the type has changed to List?

File: Singer.API/Services/UserService.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+         }
+         catch
+         {
+            throw new BadInputException();
+         }
+
+         //Ensure client is not trying to change the ID
+         if (user.Id != id)
+         {
+            throw new BadInputException();
```
Comment (wim07101993) [inline Singer.API/Services/UserService.cs]: Can't we just ignore the id of the input user?

File: Singer.API/Services/UserService.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      {
+         User userToUpdate;
+         try
+         {
+            //Check if id exists
+            userToUpdate = _appContext.Users.Single(u => u.Id == id);
+         }
+         catch
+         {
+            throw new BadInputException();
```
Comment (wim07101993) [inline Singer.API/Services/UserService.cs:97]: Maybe add some data to the exception to make it more usefull when returning an error code?

File: Singer.API/Services/UserService.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+         }
+
+         _appContext.Users.Remove(dbUser);
+         await _appContext.SaveChangesAsync();
+      }
+
+      private static Expression<Func<CareUser, bool>> Filter(string filter)
```
Comment (wim07101993) [inline Singer.API/Services/UserService.cs:127]: Is there a specific reason not to use the extension method IQueryableExtensions.Filter?

File: Tests/ServiceTests/StringFilterTests.cs [hunk 5]
```diff
@@ -0,0 +1,240 @@
+//using System;
```
Comment (wim07101993) [inline Tests/ServiceTests/StringFilterTests.cs]: A complete class in commentary?

File: Singer.API/Services/MockUserService.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      #region CONSTRUCTOR
+
+      #endregion CONSTRUCTOR
+
+
+      #region METHODS
+
+      public async Task<T> CreateUserAsync<T>(T createUser) where T : CareUser
+      {
+         // TODO User UserManager
```
Comment (wim07101993) [inline Singer.API/Services/MockUserService.cs:83]: Should this TODO be solved before merging?

File: Singer.API/Services/UserService.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      private ApplicationDbContext _appContext;
+
+      public UserService(ApplicationDbContext appContext)
+      {
+         _appContext = appContext;
+      }
+
+      public async Task<T> CreateUserAsync<T>(T createUser) where T : CareUser
+      {
+         // TODO Use Usermanager
```
Comment (wim07101993) [inline Singer.API/Services/UserService.cs:28]: Should this Todo be solved before merging?

File: Singer.API/Services/UserService.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      {
+         User dbUser;
+         try
+         {
+            //Check if id exists
+            dbUser = _appContext.Users.Single(u => u.Id == id);
+         }
+         catch
+         {
+            throw new BadInputException();
```
Comment (wim07101993) [inline Singer.API/Services/UserService.cs:120]: Again maybe a usefull error message?

File: Singer.API/Models/SearchResults.cs (hunk 1 above)
Comment (BerendWouters) [reply Singer.API/Models/SearchResults.cs:23]: The `PaginationDTO` contains an IReadonlyList too, doesn't it?

File: Singer.API/Services/UserService.cs (hunk 2 above)
Comment (BerendWouters) [reply Singer.API/Services/UserService.cs]: Agreed.

File: Singer.API/Services/UserService.cs (hunk 3 above)
Comment (BerendWouters) [reply Singer.API/Services/UserService.cs:97]: We might want to introduce extended error handling indeed, but better in a separate task.

File: Singer.API/Services/UserService.cs (hunk 4 above)
Comment (BerendWouters) [reply Singer.API/Services/UserService.cs:127]: Yes, sadly I couldn't get the Filter working with a type with subtype properties (like the `CareUser` has got the `User`)

File: Tests/ServiceTests/StringFilterTests.cs (hunk 5 above)
Comment (BerendWouters) [reply Tests/ServiceTests/StringFilterTests.cs]: I will remove the comments, the unit test is still valid and we might want to use the `Filter` object in the near future.

File: Singer.API/Services/MockUserService.cs (hunk 6 above)
Comment (BerendWouters) [reply Singer.API/Services/MockUserService.cs:83]: Yes and no. I took the easy option in this case, and just let the implementation happen when the actual ticket is implemented.

File: Singer.API/Services/UserService.cs (hunk 7 above)
Comment (BerendWouters) [reply Singer.API/Services/UserService.cs:28]: Same as above.

File: Singer.API/Models/SearchResults.cs (hunk 1 above)
Comment (wim07101993) [reply Singer.API/Models/SearchResults.cs:23]: Looked wrong, my bad

**Pull request merged**